src/eval.c

```diff
@@ -3665,6 +3673,350 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+
+    if (!eval1(&to_eval_current, &result, evaluate))
+    {
+	ga_clear(current);
+	return NULL;
+    }
+    ga_clear(current);
+
+    return result.vval.v_string;
+}
+
+/*
+ * Consumes current to evaluate,
+ * and joins its result into result.
+ *
+ * Please also see eval_template_current().
+ */
+    static int
+eval_template_current_into_result(
+	garray_T *current,
+	garray_T *result,
+	char_u quote,
+	int evaluate)
+{
+    char_u  *current_result;
+
+    current_result = eval_template_current(current, evaluate);
+    if (current_result == NULL)
+	return FAIL;
+
+    ga_concat(result, current_result);
+    vim_free(current_result);
+
+    return OK;
+}
+
+/*
+ * Allocate a variable for $"${expr}" and $'${expr}' constant.
+ * Return OK or FAIL.
+ */
+    static int
+get_template_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+{
+    garray_T	result;  // a string that is not a template string
+    garray_T	current;
+    char_u	*to_eval_current;
+    typval_T	current_result;
+    char_u	*arg_orig = *arg;
+    char_u	quote = *(*arg + 1);
+    int		is_literal_string = (quote == '\'');
+    int		is_skipping_needed;
+
+    *arg += 2;  // $'
+
+    ga_init2(&result, 1, 80);
+
+    ga_init2(&current, 1, 80);
+    ga_append(&current, quote);
+
+    // Continue while it is not NULL and it is not a closing quote.
+    for (; (**arg != NUL) &&
+	    !is_closing_quote(*arg, is_literal_string, &is_skipping_needed);
+	    ++*arg)
+    {
+	if (is_skipping_needed)
+	{
+	    ga_concatn(&current, *arg, 2);
+	    ++*arg;
+	    continue;
+	}
+	else if (**arg == '$' && *(*arg + 1) == '{')
+	{
+	    int is_eval_expr_success;
+	    int is_eval_current_success;
+
+	    ga_append(&current, quote);  // close quotation
+	    is_eval_current_success = eval_template_current_into_result(
+		    &current,
+		    &result,
+		    quote,
+		    evaluate);
+	    ga_append(&current, quote);  // open quotation
+
+	    if (!is_eval_current_success)
+	    {
+		ga_clear(&current);
+		ga_clear(&result);
+		return FAIL;
+	    }
+
+	    is_eval_expr_success =
+		eval_next_template_expr_into_result(
+			&result,
+			arg,
+			is_literal_string,
+			evaluate);
+	    if (!is_eval_expr_success)
+	    {
+		ga_clear(&current);
+		ga_clear(&result);
+		return FAIL;
+	    }
+
+	    continue;
+	}
+	ga_append(&current, (int) **arg);
+    }
+
+    /*
+     * Return FAIL if this template string is missing closing quote.
+     */
+    if (**arg == NUL)
+    {
+	ga_clear(&current);
+	ga_clear(&result);
+	semsg(_("E115: Missing quote: %s"), arg_orig);
+	return FAIL;
+    }
+    MB_PTR_ADV(*arg);
+
+    if (!evaluate)
+    {
+	ga_clear(&current);
+	ga_clear(&result);
+	return OK;
+    }
+
+    /*
+     * Consume the last current.
+     */
+    ga_append(&current, quote);  // closing quotation
+    to_eval_current = (char_u *) current.ga_data;
+    if (!eval1(&to_eval_current, &current_result, evaluate))
+    {
+	ga_clear(&current);
+	ga_clear(&result);
+	clear_tv(&current_result);
```

Comment:
Is this line needed? I think that `current_result` is not set, when `eval1()` returns `FAIL`.
Or, it might be safer to initialize `current_result` with `init_tv(&current_result);` at the top of this function.

Reply:
I don't know about `init_tv()`, but I think it should be used if it is safer than removing this line.
Should I use `init_tv()`?